packages/xo-web/src/xo-app/vm-import/esxi/esxi-import.js

```diff
@@ -0,0 +1,260 @@
+import { InputCol, LabelCol, Row } from 'form-grid'
+import { Password, Select } from 'form'
+import { SelectNetwork, SelectPool, SelectSr } from 'select-objects'
+
+import VmData from './vm-data'
+
+const getRedirectUrl = vmId => `/vms/${vmId}`
+
+@injectIntl
+@connectStore({
+  hostsById: createGetObjectsOfType('host'),
+  pifsById: createGetObjectsOfType('PIF'),
+})
+class EsxiImport extends Component {
+  state = {
+    hasCertificate: true,
+    hostIp: '',
+    isConnected: false,
+    password: '',
+    user: '',
+  }
+
+  _getDefaultNetwork = createSelector(
+    () => this.state.pool?.master,
+    () => this.props.hostsById,
+    () => this.props.pifsById,
+    (master, hostsById, pifsById) =>
+      master === undefined ? undefined : find(pick(pifsById, hostsById[master].$PIFs), pif => pif.management)?.$network
+  )
+
+  _getSelectVmOptions = createSelector(
+    () => this.state.vmsById,
+    vmsById =>
+      map(vmsById, vm => ({
+        label: vm.nameLabel,
+        value: vm.id,
+      }))
+  )
+
+  _getNetworkPredicate = createSelector(
+    () => this.props.pool?.id,
+    poolId => (poolId === undefined ? undefined : network => network.$poolId === poolId)
+  )
+
+  _getSrPredicate = createSelector(
+    () => this.props.pool?.id,
+    poolId => (poolId === undefined ? undefined : sr => isSrWritable(sr) && sr.$poolId === poolId)
+  )
+
+  _importVm = async () => {
+    const { hasCertificate, hostIp, network, password, sr, user, vm } = this.state
+    return importVmFromEsxi({
+      host: hostIp,
+      network: network?.id ?? this._getDefaultNetwork(),
+      password,
+      sr,
+      sslVerify: hasCertificate,
+      user,
+      vm: vm.value,
+    })
+  }
+
+  _connect = async () => {
+    const { hostIp, hasCertificate, password, user } = this.state
+    const vms = await esxiConnect(hostIp, user, password, hasCertificate)
+    this.setState({ isConnected: true, vmsById: keyBy(vms, 'id') })
+  }
+
+  _onChangeVm = vm => {
+    this.setState({ vm })
+  }
+
+  _onChangeNetwork = network => {
+    this.setState({ network })
+  }
+
+  _onChangePool = pool => {
+    this.setState({ pool, sr: pool.default_SR })
+  }
+  _onChangeSr = sr => {
+    this.setState({ sr })
+  }
+
+  _toggleCertificateCheck = ({ target: { checked, name } }) => {
+    this.setState({ [name]: checked })
+  }
+
+  _reset = () => {
+    this.setState({
+      hasCertificate: true,
+      hostIp: '',
+      isConnected: false,
+      network: undefined,
+      password: '',
+      pool: undefined,
+      sr: undefined,
+      user: '',
+      vm: undefined,
+      vmsById: undefined,
+    })
+  }
+
+  render() {
+    const { intl } = this.props
+    const {
+      hasCertificate,
+      hostIp,
+      isConnected,
+      network = this._getDefaultNetwork(),
+      password,
+      pool,
+      sr,
+      user,
+      vm,
+      vmsById,
+    } = this.state
+
+    return (
+      <div>
+        {!isConnected ? (
```

Comment:
Since it's almost the whole component, I think it would be more readable to write it like this:
```js
if (!isConnected) {
  return ...
}

return ...
```